File: Tableware-Android/src/uk/ac/horizon/tableware/ComponentFinder.java
package uk.ac.horizon.tableware;

import java.util.Arrays;
import org.opencv.core.Mat;

class BoundingBox
{
    public int x1, y1, x2, y2;
    public BoundingBox(int x1, int x2, int y1, int y2)
    {
        this.x1 = x1;
        this.x2 = x2;
        this.y1 = y1;
        this.y2 = y2;
    }
}

class ComponentFinder
{
    private int[] objectLabels;
    private int[] map;
    private int[] reMap;
    private int objectCount = 0;
    private BoundingBox[] objectBoundingBoxes;
    private byte[] objectClassifications;
    
    public int[] getObjectLabels(){
    	return objectLabels;
    }
     
    public int getObjectCount(){
    	return objectCount;
    }

    public BoundingBox[] getObjectBoundingBoxes(){
        return objectBoundingBoxes;
    }

    public byte[] getObjectClassifications(){
        return objectClassifications;
    }

    // Apply a connected components algorithm to the thresholded image
    public Boolean FindBlobs(Mat imgMat)
    {
        //size of matrix in terms of numbers of cols and rows.
    	int imageWidth = imgMat.width(), imageHeight = imgMat.height();
        
        // objectLabels will eventually contain the component labels for every pixel
        if (objectLabels == null || objectLabels.length != imageWidth * imageHeight)
            objectLabels = new int[imageWidth * imageHeight];
        else
        	//initialise array from previous iteration. 
            Arrays.fill(objectLabels, 0);

        int labelsCount = 0;

        // Map is used to merge regions during the second pass.
        int maxObjects = ((imageWidth / 2) + 1) * ((imageHeight / 2) + 1) + 1;
        if (map == null || map.length != maxObjects)
            map = new int[maxObjects];

        // Self map all regions at the start
        for (int i = 0; i < maxObjects; i++)
        {
            map[i] = i;
        }

        // Bounding boxes
        if (objectBoundingBoxes == null || objectBoundingBoxes.length != maxObjects)
        {
            objectBoundingBoxes = new BoundingBox[maxObjects];
        }

        // Initialise bounding boxes to extremities
        int maxValue = Integer.MAX_VALUE, minValue = Integer.MIN_VALUE;
        for (int i = 0; i < maxObjects; i++)
        {
            objectBoundingBoxes[i] = new BoundingBox(maxValue, minValue, maxValue, minValue);
        }

        // Classifications
        if (objectClassifications == null || objectClassifications.length != maxObjects)
        {
            objectClassifications = new byte[maxObjects];
        }

        int pixIndex = 0;
        int row = 0;
        int col = 0;
        
        // 1 - for pixels of the first row
        // First pixel at 0,0
        objectLabels[pixIndex] = ++labelsCount;
        ++pixIndex;

        // Process the rest of the first row. In first row, it starts scanning from the second column.
        for (col = 1; col < imageWidth; col++, pixIndex++)
        {
            // check if the previous pixel has the same binary value
            if (imgMat.get(row, col - 1)[0] == imgMat.get(row, col)[0])
            {
                // label current pixel, as the previous
                objectLabels[pixIndex] = objectLabels[pixIndex - 1];
            }
            else
            {
                // create new label
                objectLabels[pixIndex] = ++labelsCount;
            }
        }
        
        // 2 - for other rows
        // for each row
        //for (row = 1; row < imageHeight ; row++)
        long timebefore = System.currentTimeMillis();
        
        for (row = 1; row < imageHeight ; row++)
        {
            
        	// Efficiency escape
        	// for the first pixel of the row, we need to check only upper and upper-right pixels
        	//set to first column
        	col = 0;
            //First check with upper pixel.
        	if (imgMat.get(row -1, col)[0] == imgMat.get(row, col)[0])
            {
                // label current pixel, as the above
                objectLabels[pixIndex] = objectLabels[pixIndex - imageWidth];
            }
            //Otherwise check with upper right pixel.
        	else if (imgMat.get(row -1 , col)[0] == imgMat.get(row, col)[0])
            {
                // label current pixel, as the above right
                objectLabels[pixIndex] = objectLabels[pixIndex + 1 - imageWidth];
            }
        	//if none of them are equal then assign a new label. 
            else
            {
                // create new label
                objectLabels[pixIndex] = ++labelsCount;
            }
            
        	++pixIndex;

            // check left, upper left, upper and upper right pixels for the rest of the pixels.
        	//Iterate from second to second last column.
            for (col = 1; col < imageWidth - 1; col++, pixIndex++)
            {
                // check surrounding pixels
            	// check if the left pixel has the same binary value
                if (imgMat.get(row, col - 1)[0] == imgMat.get(row, col)[0]){
                    // label current pixel, as the left pixel.
                    objectLabels[pixIndex] = objectLabels[pixIndex - 1];
                }
                //otherwise check if the upper left pixel has same binary value.  
                else if (imgMat.get(row -1 , col - 1)[0] == imgMat.get(row, col)[0]){
                	objectLabels[pixIndex] = objectLabels[pixIndex - 1 - imageWidth];
                }
                //otherwise check if the upper pixel has same binary value.  
                else if (imgMat.get(row -1 , col)[0] == imgMat.get(row, col)[0]){
                	objectLabels[pixIndex] = objectLabels[pixIndex - imageWidth];
                }
                //check upper right pixel. 
                if (imgMat.get(row -1, col + 1)[0] == imgMat.get(row, col)[0]) 
                {
                    //if current pixel does not have a label.
                	if (objectLabels[pixIndex] == 0)
                    {
                        // label current pixel, as the upper right
                        objectLabels[pixIndex] = objectLabels[pixIndex + 1 - imageWidth];
                    }
                    //if current pixel has already a label.
                	else
                    {
                        int currentPixLabel = objectLabels[pixIndex];
                        int upperPixLabel = objectLabels[pixIndex + 1 - imageWidth];

                        if ((currentPixLabel != upperPixLabel) && (map[currentPixLabel] != map[upperPixLabel]))
                        {
                            // merge
                            if (map[currentPixLabel] == currentPixLabel)
                            {
                                // map current value to the upper right
                                map[currentPixLabel] = map[upperPixLabel];
                            }
                            else if (map[upperPixLabel] == upperPixLabel)
                            {
                                // map right value to the current
                                map[upperPixLabel] = map[currentPixLabel];
                            }
                            else
                            {
                                // both values are already mapped
                                map[map[currentPixLabel]] = map[upperPixLabel];
                                map[currentPixLabel] = map[upperPixLabel];
                            }

                            // reindex
                            for (int i = 1; i <= labelsCount; i++)
                            {
                                if (map[i] != i)
                                {
                                    // reindex
                                    int j = map[i];
                                    while (j != map[j])
                                    {
                                        j = map[j];
                                    }
                                    map[i] = j;
                                }
                            }
                        }
                    }
                }

                // label the object if it is not yet
                if (objectLabels[pixIndex] == 0)
                {
                    // create new label
                    objectLabels[pixIndex] = ++labelsCount;
                }
            }

            // for the last pixel of the row, we need to check left, upper and upper-left pixels.
            //check left pixel.
            if (imgMat.get(row, col - 1)[0] == imgMat.get(row, col)[0])
            {
                // label current pixel, as the left
                objectLabels[pixIndex] = objectLabels[pixIndex - 1];
            }
            //check upper left pixel.
            else if (imgMat.get(row - 1, col - 1)[0] == imgMat.get(row, col)[0])
            {
                // label current pixel, as the above left
                objectLabels[pixIndex] = objectLabels[pixIndex - 1 - imageWidth];
            }
            //check upper pixel.
            else if (imgMat.get(row - 1, col)[0] == imgMat.get(row, col)[0])
            {
                // label current pixel, as the above
                objectLabels[pixIndex] = objectLabels[pixIndex - imageWidth];
            }
            else
            {
                // create new label
                objectLabels[pixIndex] = ++labelsCount;
            }
            
            ++pixIndex;
        }
        
        long timeafter = System.currentTimeMillis();
        
        long totaltime = timeafter - timebefore;
        System.out.println(totaltime);
        // allocate remapping array
        if (reMap == null || reMap.length != map.length)
            reMap = new int[map.length];

        // count objects and prepare remapping array
        int objectsCount = 0;
        for (int i = 1; i <= labelsCount; i++)
        {
            if (map[i] == i)
            {
                // increase objects count
                reMap[i] = ++objectsCount;
            }
        }
        // second pass to complete remapping
        for (int i = 1; i <= labelsCount; i++)
        {
            if (map[i] != i)
            {
                reMap[i] = reMap[map[i]];
            }
        }

        // repair object labels
        int currentLabel = -1;
        for (row = 0; row < imageHeight; row++)
        {
            int rowIndex = row * imageWidth;
            for (col = 0; col < imageWidth; col++)
            {
                int positionIndex = rowIndex + col;
                int regionID = reMap[objectLabels[positionIndex]];
                objectLabels[positionIndex] = regionID;

                // Check classification
                if (regionID != currentLabel)
                {
                    objectClassifications[regionID] = imgMat.get(row, col)[0] > 0 ? (byte)1 : (byte)0;
                    currentLabel = regionID;
                }

                // Check and update bounding box 
                // x
                if (col < objectBoundingBoxes[regionID].x1)
                    objectBoundingBoxes[regionID].x1 = col;
                if (col > objectBoundingBoxes[regionID].x2)
                    objectBoundingBoxes[regionID].x2 = col;
                // y
                if (row < objectBoundingBoxes[regionID].y1)
                    objectBoundingBoxes[regionID].y1 = row;
                if (row > objectBoundingBoxes[regionID].y2)
                    objectBoundingBoxes[regionID].y2 = row;
            }
        }
        
        // Labels from 0 - max inclusive
        objectCount = objectsCount + 1;
        return true;
    }
}	

File: windows/Ceramics/CeramicsBase/ComponentFinder.cs
/*
    Copyright (c) 2011 University of Nottingham.
    Contact <richard.mortier@nottingham.ac.uk> for more info.
    Original code by Michael Pound <mpp@cs.nott.ac.uk>.

    This file is part of Ceramics.

    Ceramics is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    Ceramics is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public
    License along with Ceramics.  If not, see
    <http://www.gnu.org/licenses/>.
*/
﻿using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;
using System.Drawing.Imaging;

using AForge;
using AForge.Imaging;

namespace Ceramics
{
    struct BoundingBox
    {
        public int x1, y1, x2, y2;
        public BoundingBox(int x1, int x2, int y1, int y2)
        {
            this.x1 = x1;
            this.x2 = x2;
            this.y1 = y1;
            this.y2 = y2;
        }
    }
    class ComponentFinder
    {
        private int[] objectLabels;
        public int[] ObjectLabels
        {
            get
            {
                return objectLabels;
            }
        }

        private int[] map;
        private int[] reMap;

        private int objectCount = 0;
        public int ObjectCount
        {
            get
            {
                return objectCount;
            }
        }

        private BoundingBox[] objectBoundingBoxes;
        public BoundingBox[] ObjectBoundingBoxes
        {
            get
            {
                return objectBoundingBoxes;
            }
        }

        private byte[] objectClassifications;
        public byte[] ObjectClassifications
        {
            get
            {
                return objectClassifications;
            }
        }

        public ComponentFinder()
        {
        }

        public unsafe bool FindBlobs(UnmanagedImage grayBufferUMI)
        {
                if (grayBufferUMI.PixelFormat != PixelFormat.Format8bppIndexed)
                    return false;

                // Apply a connected components algorithm to the thresholded image
                int imageWidth = grayBufferUMI.Width, imageHeight = grayBufferUMI.Height;
                int stride = grayBufferUMI.Stride;

                // objectLabels will eventually contain the component labels for every pixel
                if (objectLabels == null || objectLabels.Length != imageWidth * imageHeight)
                    objectLabels = new int[imageWidth * imageHeight];
                else
                    Array.Clear(objectLabels, 0, objectLabels.Length);

                int labelsCount = 0;

                // Map is used to merge regions during the second pass.
                int maxObjects = ((imageWidth / 2) + 1) * ((imageHeight / 2) + 1) + 1;
                if (map == null || map.Length != maxObjects)
                    map = new int[maxObjects];

                // Self map all regions at the start
                for (int i = 0; i < maxObjects; i++)
                {
                    map[i] = i;
                }

                // Bounding boxes
                if (objectBoundingBoxes == null || objectBoundingBoxes.Length != maxObjects)
                {
                    objectBoundingBoxes = new BoundingBox[maxObjects];
                }

                // Initialise bounding boxes to extremities
                int maxValue = Int32.MaxValue, minValue = Int32.MinValue;
                for (int i = 0; i < maxObjects; i++)
                {
                    objectBoundingBoxes[i] = new BoundingBox(maxValue, minValue, maxValue, minValue);
                }

                // Classifications
                if (objectClassifications == null || objectClassifications.Length != maxObjects)
                {
                    objectClassifications = new byte[maxObjects];
                }

                // Pointer to image data
                byte* src = (byte*)grayBufferUMI.ImageData.ToPointer();
                int p = 0;
                int offset = stride - imageWidth;

                // 1 - for pixels of the first row
                // First pixel at 0,0
                objectLabels[p] = ++labelsCount;
                ++src;
                ++p;

                // Process the rest of the first row
                for (int x = 1; x < imageWidth; x++, src++, p++)
                {
                    // check if the previous pixel is the same binary value
                    if (src[-1] == src[0])
                    {
                        // label current pixel, as the previous
                        objectLabels[p] = objectLabels[p - 1];
                    }
                    else
                    {
                        // create new label
                        objectLabels[p] = ++labelsCount;
                    }
                }
                src += offset;

                // 2 - for other rows
                // for each row
                for (int y = 1; y < imageHeight; y++)
                {
                    // Efficiency escape

                    // for the first pixel of the row, we need to check
                    // only upper and upper-right pixels
                    // check surrounding pixels
                    if (src[-stride] == src[0])
                    {
                        // label current pixel, as the above
                        objectLabels[p] = objectLabels[p - imageWidth];
                    }
                    else if (src[1 - stride] == src[0])
                    {
                        // label current pixel, as the above right
                        objectLabels[p] = objectLabels[p + 1 - imageWidth];
                    }
                    else
                    {
                        // create new label
                        objectLabels[p] = ++labelsCount;
                    }
                    ++src;
                    ++p;

                    // check left pixel and three upper pixels for the rest of pixels
                    for (int x = 1; x < imageWidth - 1; x++, src++, p++)
                    {
                        // check surrounding pixels
                        if (src[-1] == src[0])
                        {
                            // label current pixel, as the left
                            objectLabels[p] = objectLabels[p - 1];
                        }
                        else if (src[-1 - stride] == src[0])
                        {
                            // label current pixel, as the above left
                            objectLabels[p] = objectLabels[p - 1 - imageWidth];
                        }
                        else if (src[-stride] == src[0])
                        {
                            // label current pixel, as the above
                            objectLabels[p] = objectLabels[p - imageWidth];
                        }

                        if (src[1 - stride] == src[0])
                        {
                            if (objectLabels[p] == 0)
                            {
                                // label current pixel, as the above right
                                objectLabels[p] = objectLabels[p + 1 - imageWidth];
                            }
                            else
                            {
                                int l1 = objectLabels[p];
                                int l2 = objectLabels[p + 1 - imageWidth];

                                if ((l1 != l2) && (map[l1] != map[l2]))
                                {
                                    // merge
                                    if (map[l1] == l1)
                                    {
                                        // map left value to the right
                                        map[l1] = map[l2];
                                    }
                                    else if (map[l2] == l2)
                                    {
                                        // map right value to the left
                                        map[l2] = map[l1];
                                    }
                                    else
                                    {
                                        // both values already mapped
                                        map[map[l1]] = map[l2];
                                        map[l1] = map[l2];
                                    }

                                    // reindex
                                    for (int i = 1; i <= labelsCount; i++)
                                    {
                                        if (map[i] != i)
                                        {
                                            // reindex
                                            int j = map[i];
                                            while (j != map[j])
                                            {
                                                j = map[j];
                                            }
                                            map[i] = j;
                                        }
                                    }
                                }
                            }
                        }

                        // label the object if it is not yet
                        if (objectLabels[p] == 0)
                        {
                            // create new label
                            objectLabels[p] = ++labelsCount;
                        }
                    }

                    // for the last pixel of the row, we need to check
                    // only upper and upper-left pixels
                    // check surrounding pixels
                    if (src[-1] == src[0])
                    {
                        // label current pixel, as the left
                        objectLabels[p] = objectLabels[p - 1];
                    }
                    else if (src[-1 - stride] == src[0])
                    {
                        // label current pixel, as the above left
                        objectLabels[p] = objectLabels[p - 1 - imageWidth];
                    }
                    else if (src[-stride] == src[0])
                    {
                        // label current pixel, as the above
                        objectLabels[p] = objectLabels[p - imageWidth];
                    }
                    else
                    {
                        // create new label
                        objectLabels[p] = ++labelsCount;
                    }
                    ++src;
                    ++p;

                    src += offset;
                }

                // allocate remapping array
                if (reMap == null || reMap.Length != map.Length)
                    reMap = new int[map.Length];

                // count objects and prepare remapping array
                int objectsCount = 0;
                for (int i = 1; i <= labelsCount; i++)
                {
                    if (map[i] == i)
                    {
                        // increase objects count
                        reMap[i] = ++objectsCount;
                    }
                }
                // second pass to complete remapping
                for (int i = 1; i <= labelsCount; i++)
                {
                    if (map[i] != i)
                    {
                        reMap[i] = reMap[map[i]];
                    }
                }

                // repair object labels
                int currentLabel = -1;
                src = (byte*)grayBufferUMI.ImageData.ToPointer();
                for (int y = 0; y < imageHeight; y++)
                {
                    int rowIndex = y * imageWidth;
                    for (int x = 0; x < imageWidth; x++)
                    {
                        int positionIndex = rowIndex + x;
                        int regionID = reMap[objectLabels[positionIndex]];
                        objectLabels[positionIndex] = regionID;

                        // Check classification
                        if (regionID != currentLabel)
                        {
                            objectClassifications[regionID] = src[y * stride + x] > 0 ? (byte)1 : (byte)0;
                            currentLabel = regionID;
                        }

                        // Check and update bounding box 
                        // x
                        if (x < objectBoundingBoxes[regionID].x1)
                            objectBoundingBoxes[regionID].x1 = x;
                        if (x > objectBoundingBoxes[regionID].x2)
                            objectBoundingBoxes[regionID].x2 = x;
                        // y
                        if (y < objectBoundingBoxes[regionID].y1)
                            objectBoundingBoxes[regionID].y1 = y;
                        if (y > objectBoundingBoxes[regionID].y2)
                            objectBoundingBoxes[regionID].y2 = y;
                    }
                }
                // Labels from 0 - max inclusive
                objectCount = objectsCount + 1;
                return true;
        }
        
    }
}
